### PR TITLE
Spring 05

### PR DESCRIPTION
**Pull request by engineer027**

### CODE REVIEW

**Review by Ilya-brinchuk — APPROVED** [review]

File: src/main/java/com/dev/cinema/controllers/MovieSessionController.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping(" /movie-sessions")
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controllers/MovieSessionController.java]:
```suggestion
@RequestMapping("/movie-sessions")
```

File: src/main/java/com/dev/cinema/controllers/MovieSessionController.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+                                  MovieSessionMapper movieSessionMapper) {
+        this.movieSessionService = movieSessionService;
+        this.movieSessionMapper = movieSessionMapper;
+    }
+
+    @GetMapping("/available")
+    public List<MovieSessionResponseDto> findAvailableSessions(@RequestParam Long movieId,
+                                                               @RequestParam
+                                                    @DateTimeFormat(pattern = "dd.MM.yyyy")
+                                                            LocalDate date) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controllers/MovieSessionController.java]:
```suggestion
    public List<MovieSessionResponseDto> findAvailableSessions(
                                      @RequestParam Long movieId,
                                      @RequestParam @DateTimeFormat(pattern = "dd.MM.yyyy")
                                      LocalDate date) {
```
let's try to make this more readable

File: src/main/java/com/dev/cinema/controllers/MovieSessionController.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    @PostMapping
+    public void create(@RequestBody MovieSessionRequestDto movieSessionRequestDto) {
+        MovieSession movieSession = movieSessionMapper
+                .mapRequestDtoToMovieSession(movieSessionRequestDto);
+        movieSessionService.add(movieSession);
+    }
+
+    @DeleteMapping
+    public void delete(@RequestBody MovieSessionDeleteDto movieSessionDeleteDto) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controllers/MovieSessionController.java]: please see this thread - https://mateacademy.slack.com/archives/G01EHB8CWAE/p1613330083125300

File: src/main/java/com/dev/cinema/dao/MovieSessionDao.java [hunk 4]
```diff
@@ -3,10 +3,17 @@
...
 
 public interface MovieSessionDao {
 
     MovieSession add(MovieSession movieSession);
 
     List<MovieSession> findAvailableSessions(Long movieId, LocalDate date);
+
+    MovieSession update(MovieSession movieSession);
+
+    MovieSession delete(MovieSession movieSession);
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/dao/MovieSessionDao.java]:
```suggestion
    void delete(Long movieSessionId);
```

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java [hunk 5]
```diff
@@ -60,4 +61,59 @@ public MovieSession add(MovieSession movieSession) {
...
+    }
+
+    @Override
+    public MovieSession delete(MovieSession movieSession) {
+        Transaction transaction = null;
+        Session session = null;
+        try {
+            session = sessionFactory.openSession();
+            transaction = session.beginTransaction();
+            session.delete(movieSession);
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java:96]: let's use load() method to get the session by id before deleting it

File: src/main/java/com/dev/cinema/model/dto/MovieSessionDeleteDto.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
+package com.dev.cinema.model.dto;
+
+import java.time.LocalDateTime;
+
+public class MovieSessionDeleteDto {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/model/dto/MovieSessionDeleteDto.java]: let's remove this

File: src/main/java/com/dev/cinema/model/dto/MovieSessionRequestDto.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
+package com.dev.cinema.model.dto;
+
+import java.time.LocalDateTime;
+
+public class MovieSessionRequestDto {
+    private LocalDateTime showTime;
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/model/dto/MovieSessionRequestDto.java]: how can you pass LocalDateTime in json?

File: src/main/java/com/dev/cinema/model/dto/MovieSessionResponseDto.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
+package com.dev.cinema.model.dto;
+
+import com.dev.cinema.model.Movie;
+
+public class MovieSessionResponseDto {
+    private Long id;
+    private Movie movie;
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/model/dto/MovieSessionResponseDto.java]: let's not fill the Dto with entity - use movieId instead

File: src/main/java/com/dev/cinema/model/dto/MovieSessionResponseDto.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    private String movieTitle;
+    private Long cinemaHallId;
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/model/dto/MovieSessionResponseDto.java:7]: what about showTime?

File: src/main/java/com/dev/cinema/model/dto/MovieSessionUpdateDto.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
+package com.dev.cinema.model.dto;
+
+import java.time.LocalDateTime;
+
+public class MovieSessionUpdateDto {
+    private Long id;
+    private LocalDateTime showTime;
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/model/dto/MovieSessionUpdateDto.java]: how can you pass LocalDateTime in json?

File: src/main/java/com/dev/cinema/service/mapper/CinemaHallMapperImpl.java [hunk 11]
```diff
@@ -0,0 +1,26 @@
+package com.dev.cinema.service.mapper;
+
+import com.dev.cinema.model.CinemaHall;
+import com.dev.cinema.model.dto.CinemaHallRequestDto;
+import com.dev.cinema.model.dto.CinemaHallResponseDto;
+import org.springframework.stereotype.Component;
+
+@Component
+public class CinemaHallMapperImpl implements CinemaHallMapper {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/service/mapper/CinemaHallMapperImpl.java]: we don't really need interfaces for these mappers - let's remove the interfaces
```suggestion
public class CinemaHallMapper {
```

File: src/main/java/com/dev/cinema/service/mapper/CinemaHallMapperImpl.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+    @Override
+    public CinemaHallResponseDto mapCinemaHallToDto(CinemaHall cinemaHall) {
+        CinemaHallResponseDto cinemaHallResponseDto = new CinemaHallResponseDto();
+        cinemaHallResponseDto.setDescription(cinemaHall.getDescription());
+        cinemaHallResponseDto.setId(cinemaHall.getId());
+        cinemaHallResponseDto.setCapacity(cinemaHallResponseDto.getCapacity());
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/service/mapper/CinemaHallMapperImpl.java]: is this line ok?

File: src/main/java/com/dev/cinema/service/mapper/MovieSessionMapperImpl.java [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        movieSession.setId(movieSessionUpdateDto.getId());
+        movieSession.setShowTime(movieSessionUpdateDto.getShowTime());
+        movieSession.setMovie(movieService.get(movieSessionUpdateDto.getMovieId()));
+        movieSession.setCinemaHall(cinemaHallService.get(movieSessionUpdateDto
+                .getCinemaHallId()));
+        return movieSession;
+    }
+
+    @Override
+    public MovieSession mapDeleteDtoToMovieSession(MovieSessionDeleteDto movieSessionDeleteDto) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/service/mapper/MovieSessionMapperImpl.java]: let's remove this method

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/dev/cinema/controllers/MovieSessionController.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+    public void create(@RequestBody MovieSessionRequestDto movieSessionRequestDto) {
+        MovieSession movieSession = movieSessionMapper
+                .mapRequestDtoToMovieSession(movieSessionRequestDto);
+        movieSessionService.add(movieSession);
+    }
+
+    @DeleteMapping("/{movieSessionId}")
+    public void delete(@PathVariable Long movieSessionId) {
+        movieSessionService.delete(movieSessionId);
+
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controllers/MovieSessionController.java]: redundant empty line

File: src/main/java/com/dev/cinema/service/mapper/CinemaHallMapperImpl.java [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+import com.dev.cinema.model.dto.CinemaHallResponseDto;
+import org.springframework.stereotype.Component;
+
+@Component
+public class CinemaHallMapperImpl {
+    public CinemaHallResponseDto mapCinemaHallToDto(CinemaHall cinemaHall) {
+        CinemaHallResponseDto cinemaHallResponseDto = new CinemaHallResponseDto();
+        cinemaHallResponseDto.setDescription(cinemaHall.getDescription());
+        cinemaHallResponseDto.setId(cinemaHall.getId());
+        cinemaHallResponseDto.setCapacity(cinemaHallResponseDto.getCapacity());
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/service/mapper/CinemaHallMapperImpl.java]: again, it is something wrong with this line

File: src/main/java/com/dev/cinema/service/mapper/CinemaHallMapperImpl.java [hunk 16]
```diff
@@ -0,0 +1,24 @@
+package com.dev.cinema.service.mapper;
+
+import com.dev.cinema.model.CinemaHall;
+import com.dev.cinema.model.dto.CinemaHallRequestDto;
+import com.dev.cinema.model.dto.CinemaHallResponseDto;
+import org.springframework.stereotype.Component;
+
+@Component
+public class CinemaHallMapperImpl {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/service/mapper/CinemaHallMapperImpl.java]:
```suggestion
public class CinemaHallMapper {
```
same for all mappers

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**